src/analysis/normed_space/conformal.lean

```diff
@@ -0,0 +1,249 @@
+  {R M N G : Type*} [nondiscrete_normed_field R]
+  [normed_group M] [normed_group N] [normed_group G]
+  [normed_space R M] [normed_space R N] [normed_space R G]
+
+lemma is_conformal_map_iff (f' : E →L[ℝ] F) :
+  is_conformal_map ℝ f' ↔ ∃ (c : ℝ) (hc : 0 < c),
+  ∀ (u v : E), ⟪f' u, f' v⟫ = (c : ℝ) * ⟪u, v⟫ :=
+begin
+  split,
+  { rintros ⟨c₁, hc₁, li, h⟩,
+    refine ⟨c₁ * c₁, mul_self_pos hc₁, λ u v, _⟩,
+    simp only [h, pi.smul_apply, inner_map_map,
+               real_inner_smul_left, real_inner_smul_right, mul_assoc], },
+  { rintros ⟨c₁, hc₁, huv⟩,
+    let c := real.sqrt c₁⁻¹,
+    have hc : c ≠ 0 := λ w, by {simp only [c] at w;
+      exact (real.sqrt_ne_zero'.mpr $ inv_pos.mpr hc₁) w},
+    let f₁ := c • f',
+    have minor : (f₁ : E → F) = c • f' := rfl,
+    have minor' : (f' : E → F) = c⁻¹ • f₁ := by ext;
+      simp_rw [minor, pi.smul_apply]; rw [smul_smul, inv_mul_cancel hc, one_smul],
+    refine ⟨c⁻¹, inv_ne_zero hc, f₁.to_linear_map.isometry_of_inner (λ u v, _), minor'⟩,
+    simp_rw [to_linear_map_eq_coe, continuous_linear_map.coe_coe, minor, pi.smul_apply],
+    rw [real_inner_smul_left, real_inner_smul_right,
+        huv u v, ← mul_assoc, ← mul_assoc,
+        real.mul_self_sqrt $ le_of_lt $ inv_pos.mpr hc₁,
+        inv_mul_cancel $ ne_of_gt hc₁, one_mul], },
+end
+
+namespace is_conformal_map
+
+lemma comp {f' : M →L[R] N} {g' : N →L[R] G}
+  (hf' : is_conformal_map R f') (hg' : is_conformal_map R g') :
+  is_conformal_map R (g'.comp f') :=
+begin
+  rcases hf' with ⟨cf, hcf, lif, hlif⟩,
+  rcases hg' with ⟨cg, hcg, lig, hlig⟩,
+  refine ⟨cg * cf, mul_ne_zero hcg hcf, lig.comp lif, funext (λ x, _)⟩,
+  simp only [coe_comp', linear_isometry.coe_comp, hlif, hlig, pi.smul_apply,
+             function.comp_app, linear_isometry.map_smul, smul_smul],
+end
+
+lemma preserves_angle (f' : E →L[ℝ] F) (h : is_conformal_map ℝ f') (u v : E) :
+  inner_product_geometry.angle (f' u) (f' v) = inner_product_geometry.angle u v :=
+begin
+  repeat { rw inner_product_geometry.angle },
+  suffices new : ⟪f' u, f' v⟫ / (∥f' u∥ * ∥f' v∥) = ⟪u, v⟫ / (∥u∥ * ∥v∥),
+  { rw new, },
+  { rcases h with ⟨c₁, hc₁, li, hf'⟩,
+    have minor : ∥c₁∥ ≠ 0 := λ w, hc₁ (norm_eq_zero.mp w),
+    simp_rw [hf', pi.smul_apply],
+    rw [real_inner_smul_left, real_inner_smul_right, ← mul_assoc, inner_map_map],
+    repeat { rw [norm_smul, norm_map] },
+    rw [← mul_assoc],
+    exact calc c₁ * c₁ * ⟪u, v⟫ / (∥c₁∥ * ∥u∥ * ∥c₁∥ * ∥v∥)
+            = c₁ * c₁ * ⟪u, v⟫ / (∥c₁∥ * ∥c₁∥ * ∥u∥ * ∥v∥) : by simp only [mul_comm, mul_assoc]
+        ... = c₁ * c₁ * ⟪u, v⟫ / (abs c₁ * abs c₁ * ∥u∥ * ∥v∥) : by rw [real.norm_eq_abs]
+        ... = c₁ * c₁ * ⟪u, v⟫ / (c₁ * c₁ * ∥u∥ * ∥v∥) : by rw [← pow_two, ← sq_abs, pow_two]
+        ... = c₁ * (c₁ * ⟪u, v⟫) / (c₁ * (c₁ * (∥u∥ * ∥v∥))) : by simp only [mul_assoc]
+        ... = (c₁ * ⟪u, v⟫) / (c₁ * (∥u∥ * ∥v∥)) : by rw mul_div_mul_left _ _ hc₁
+        ... = ⟪u, v⟫ / (∥u∥ * ∥v∥) : by rw mul_div_mul_left _ _ hc₁, },
```

Comment:
```suggestion
  obtain ⟨c, hc, li, hcf⟩ := h,
  suffices : c * (c * inner u v) / (∥c∥ * ∥u∥ * (∥c∥ * ∥v∥)) = inner u v / (∥u∥ * ∥v∥),
  { simp [this, inner_product_geometry.angle, hcf, norm_smul, inner_smul_left, inner_smul_right] },
  by_cases hu : ∥u∥ = 0,
  { simp [norm_eq_zero.mp hu] },
  by_cases hv : ∥v∥ = 0,
  { simp [norm_eq_zero.mp hv] },
  have hc : ∥c∥ ≠ 0 := λ w, hc (norm_eq_zero.mp w),
  field_simp,
  have : c * c = ∥c∥ * ∥c∥ := by simp [real.norm_eq_abs, abs_mul_abs_self],
  convert congr_arg (λ x, x * ⟪u, v⟫ * ∥u∥ * ∥v∥) this using 1; ring,
```
Some tricks for this kind of calculation -- feel free to use the whole thing or just some ideas.
- establish nonzeroness of everything and then clear denominators using `field_simp`
- separate into pieces which `simp` can do and pieces which `ring` can do.